bin/server/cmd/component_omni.go

```diff
@@ -48,19 +54,70 @@ func createMetadataDB(activeProfile *Profile, logger *zap.Logger) (*metadataDB,
 }
 
 func (m *metadataDB) connect() (*store.DB, error) {
-	if err := m.pgInstance.Start(m.profile.datastorePort, os.Stderr, os.Stderr); err != nil {
+	if useEmbedDB() {
+		if err := m.pgInstance.Start(m.profile.datastorePort, os.Stderr, os.Stderr); err != nil {
+			return nil, err
+		}
+		m.pgStarted = true
+
+		// Even when Postgres opens Unix domain socket only for connection, it still requires a port as ID to differentiate different Postgres instances.
+		connCfg := dbdriver.ConnectionConfig{
+			Username:    m.profile.pgUser,
+			Password:    "",
+			Host:        common.GetPostgresSocketDir(),
+			Port:        fmt.Sprintf("%d", m.profile.datastorePort),
+			StrictUseDb: false,
+		}
+		db := store.NewDB(m.l, connCfg, m.profile.demoDataDir, readonly, version, m.profile.mode)
+		return db, nil
+	}
+
+	u, err := url.Parse(pgURL)
+	if err != nil {
 		return nil, err
 	}
-	m.pgStarted = true
 
-	// Even when Postgres opens Unix domain socket only for connection, it still requires a port as ID to differentiate different Postgres instances.
+	m.l.Info("Establishing external PostgreSQL connection...", zap.String("pgURL", u.Redacted()))
+
+	if u.Scheme != "postgres" && u.Scheme != "postgresql" {
+		return nil, fmt.Errorf("invalid connection protocol: %s", u.Scheme)
+	}
+
 	connCfg := dbdriver.ConnectionConfig{
-		Username:    m.profile.pgUser,
-		Password:    "",
-		Host:        common.GetPostgresSocketDir(),
-		Port:        fmt.Sprintf("%d", m.profile.datastorePort),
-		StrictUseDb: false,
+		StrictUseDb: true,
+	}
+
+	if u.User != nil {
+		connCfg.Username = u.User.Username()
+		connCfg.Password, _ = u.User.Password()
 	}
+
+	if connCfg.Username == "" {
+		return nil, fmt.Errorf("missing user in the --pg connection string")
+	}
+
+	if host, port, err := net.SplitHostPort(u.Host); err != nil {
+		connCfg.Host = u.Host
+	} else {
+		connCfg.Host = host
+		connCfg.Port = port
+	}
+
+	// By default, follow the PG convention to use user name as the database name
+	connCfg.Database = connCfg.Username
```

Comment:
according to the doc, user must explicitly specify the database name.

Reply:
my bad, I didn't double-check carefully after copying from slim